File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode1935MaximumNumberOfWordsYouCanType.cpp
#include "LeetCode1935MaximumNumberOfWordsYouCanType.h"

#include <unordered_set>

int LeetCode1935MaximumNumberOfWordsYouCanType::canBeTypedWords(std::string text, std::string brokenLetters)
{
    std::unordered_set<char> letters{};
    letters.reserve(brokenLetters.size());
    for (auto c : brokenLetters)
    {
        letters.insert(c);
    }
    int count = 0;
    bool hasBrokenLetter{};
    std::string current{};
    for (auto c : text)
    {
        if (c == ' ')
        {
            if (hasBrokenLetter)
            {
                hasBrokenLetter = false;
                current={};
                continue;
            }

            if (current.empty())
            {
                continue;
            }
            count++;
            current={};
            continue;
        }
        if (letters.contains(c))
        {
            hasBrokenLetter = true;
        }
        current+=c;
    }
    if (!hasBrokenLetter && !current.empty())
    {
        count++;
    }
    return count;   
}

File: LeetCodeProblemsGo/LeetCode1935MaximumNumberOfWordsYouCanType/LeetCode1935MaximumNumberOfWordsYouCanType.go
package LeetCode1935MaximumNumberOfWordsYouCanType

func canBeTypedWords(text string, brokenLetters string) int {
	letters := make(map[byte]struct{}, len(brokenLetters))
	for _, c := range brokenLetters {
		letters[byte(c)] = struct{}{}
	}
	count := 0
	hasBrokenLetter := false
	currentLength := 0
	for _, c := range text {
		if c == ' ' {
			if !hasBrokenLetter && currentLength > 0 {
				count++
				continue
			}
			currentLength = 0
			hasBrokenLetter = false
			continue
		}
		currentLength++
		if _, exists := letters[byte(c)]; exists {
			hasBrokenLetter = true
		}
	}
	if !hasBrokenLetter && currentLength > 0 {
		count++
	}
	return count
}
